src/mlpack/methods/bayesian_ridge/bayesian_ridge.cpp

```diff
@@ -0,0 +1,262 @@
+/**
+ * @file bayesian_ridge.cpp
+ * @author Clement Mercier 
+ *
+ * Implementation of Bayesian Ridge regression.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include "bayesian_ridge.hpp"
+#include <mlpack/core/util/log.hpp>
+#include <mlpack/core/util/timers.hpp>
+
+using namespace mlpack;
+using namespace mlpack::regression;
+
+
+BayesianRidge::BayesianRidge(const bool centerData,
+                             const bool scaleData,
+                             const int nIterMax,
+                             const double tol) :
+  centerData(centerData),
+  scaleData(scaleData),
+  nIterMax(nIterMax),
+  tol(tol)
+{/* Nothing to do */}
+
+double BayesianRidge::Train(const arma::mat& data,
+                            const arma::rowvec& responses)
+{
+  Timer::Start("bayesian_ridge_regression");
+
+  arma::mat phi;
+  arma::rowvec t;
+  arma::colvec eigval;
+  arma::mat eigvec;
+  arma::colvec eigvali;
+
+  // Preprocess the data. Center and scale.
+  responsesOffset = CenterScaleData(data,
+                                    responses,
+                                    centerData,
+                                    scaleData,
+                                    phi,
+                                    t,
+                                    dataOffset,
+                                    dataScale);
+
+  // Compute this quantities once and for all.
+  const arma::colvec vecphitT = phi * t.t();
+
+  // Enforce symmetry of the covariance matrix before eig_sym.
+  const arma::mat phiphiT =  arma::symmatu(phi * phi.t());
+
+  if (arma::eig_sym(eigval, eigvec, phiphiT) == false)
+  {
+    Log::Warn << "Eigen Decomposition failed as Eigen Value "
+              << "does not exists ." << std::endl;
+    return -1;
+  }
+
+  // Initialize the hyperparameters and
+  // begin with an infinitely broad prior.
+  alpha = 1e-6;
+  beta =  1 / (var(t) * 0.1);
+
+  unsigned short nIterMax = 50;
+  unsigned short i = 0;
+  double deltaAlpha = 1, deltaBeta = 1, crit = 1;
+  arma::mat matA = arma::eye<arma::mat>(data.n_rows, data.n_rows);
+
+  while ((crit > tol) && (i < nIterMax))
+  {
+    deltaAlpha = -alpha;
+    deltaBeta = -beta;
+
+    // Compute the posterior statistics.
+    // with inv()
+    matA.diag().fill(alpha);
+    // inv is used instead of solve because we need the covariance matrix to
+    // compute the prediction uncertainties. If solve is used, matCovariance
+    // must be comptuted at the end of the loop.
```

Comment:
Can you do a quick benchmark here?  I think that `solve()` will often be a faster approach, even if once we converge we need to compute `matCovariance`.  However that depends on the solution technique... if `inv_sympd()` uses, e.g., the Cholesky decomposition, I think it should be equivalent.  It's still worth trying either way, in my opinion.

Reply:
olesky. The first dataset had a size of 288 points x 292 highly correlated features (spectroscopic data) and the second was about 100 points x 10 features. The training times are : 
                     cholesky - solve - inv_sympd
Dataset 1 ms     528    -   246  -  305
Dataset 2 µs      651    -   506  -  328

In the case of solve(), the time to compute the covariance at the end of the loop is included. 
Here is the code I used. 
`// matCovariance = inv_sympd(matA + phiphiT * beta);
    // omega = (matCovariance * vecphitT) * beta;

    // // with solve()
    matA.diag().fill(alpha / beta);
    omega = solve(matA + phiphiT, vecphitT);

    // // with chol()
    // const arma::mat R = chol(matA + phiphiT * beta);
    // omega = solve(R, solve(R.t(), vecphitT));
    // omega *= beta;
    `

Reply:
It would seem that this is in favor of solve(). But from your suggestion about the re-expression of the optimization, all of this can become obsolete :)

Reply:
What do you think about computing the covariance only at the first call to the .Predict() that compute the uncertainties ?